tests/cupy_tests/cuda_tests/test_texture.py

```diff
@@ -0,0 +1,110 @@
+import unittest
+
+import numpy
+
+import cupy
+from cupy import testing
+from cupy.cuda import runtime
+from cupy.cuda.texture import (ChannelFormatDescriptor, CUDAArray,
+                               ResourceDescriptor, TextureDescriptor,
+                               TextureObject)
+
+
+stream_for_async_cpy = cupy.cuda.Stream()
+dev = cupy.cuda.Device(runtime.getDevice())
+
+
+@testing.parameterize(*testing.product({
+    'xp': (numpy, cupy),
+    'stream': (None, stream_for_async_cpy),
+    'dimensions': [(67, 0, 0), (67, 19, 0), (67, 19, 31)],
+    })
+)
+class TestCUDAArray(unittest.TestCase):
+    def test_array_gen_cpy(self):
+        xp = self.xp
+        stream = self.stream
+        width, height, depth = self.dimensions
+        dim = 3 if depth != 0 else 2 if height != 0 else 1
+
+        # generate input data and allocate output buffer
+        if dim == 3:
+            arr = xp.random.random((depth, height, width)).astype(cupy.float32) # 3D random array
+        elif dim == 2:
+            arr = xp.random.random((height, width)).astype(cupy.float32) # 2D random array
+        else:
+            arr = xp.random.random((width,)).astype(cupy.float32) # 1D random array
+        arr2 = xp.zeros_like(arr)
+
+        assert arr.flags['C_CONTIGUOUS']
+        assert arr2.flags['C_CONTIGUOUS']
+
+        # create a CUDA array
+        ch = ChannelFormatDescriptor(32, 0, 0, 0, runtime.cudaChannelFormatKindFloat)
+        cu_arr = CUDAArray(ch, width, height, depth)
+        
+        # copy from input to CUDA array, and back to output
+        cu_arr.copy_from(arr, stream)
```

Comment:
the problem seems to be in one of this two copy functions, if commented then the tests run ok

Reply:
The problem only happens when 'xp' is cupy, running the `TestCUDAArray` with `xp = numpy` doesnt make `Test2DTexture` to fail.

Reply:
> Does it matter that I didn't allocate CUDAArray from CuPy's memory pool?

Don't think that's relevant. I disabled the pool using `cupy.cuda.set_allocator(None)` and the problem persisted.

Reply:
One of the things that I saw, is that the some of the values read from the Texture in the kernel, were the values that the other test set!!!!, so there is a conflict with cudaArrays allocated memory ...

Reply:
That's really odd. I didn't see that behavior. 

What I saw so far pretty consistently is that when the `AssertionError` was raised, it's because `real_output` only got the first row correctly; the rest is all duplicate:
```python
> print(real_output)
[[0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]
 [0. 1. 2. 3. 4. 5. 6. 7.]]
```
Is this what you observed?